### DISCUSSION TITLE
Difference between the five models for variant prediction

### DISCUSSION THREAD

**Discussion by TheBil99** [facebookresearch/esm#492] [Q&A]
Hi, I see that you provided five different ESM models for variant prediction: esm1v_t33_650M_UR90S_1, esm1v_t33_650M_UR90S_2, esm1v_t33_650M_UR90S_3, esm1v_t33_650M_UR90S_4, esm1v_t33_650M_UR90S_5. Unfortunately, I was not able to find an explanation of the differences between the models, could you clarify this aspect? 
Thank you in advance!! :)

**Comment by tomsercu** [comment]
Different random seeds for model training, which were used to create the ensemble for predictions in the paper.

**Reply by TheBil99** [reply]
What does the random seed control in the model? Also, could you explain which ensembling strategy did you use? From the paper it seems you use a distinct strategy from a simple average: in table 2 you report two different rho values, one for average of the five models and one for the ensemble.

**Comment by tomsercu** [comment]
Random seed in deep learning controls the random initialization of the weights, and data shuffling.
Average: first compute the performance of each model separately and average it.
Ensemble: average the variant effect predictions across the model, then compute the metrics.
Hope that helps!

**Reply by TheBil99** [reply]
@tomsercu thank you for your answer, it's been very helpful. Just one thing, I am trying to reproduce the results obtained by you just to be sure the procedure is correct, before proceeding to apply it to my dataset of interest. I am thus trying to reproduce at least the same spearman rho you obtain with the variant prediction from the BLAT_ECOLX_Ranganathan2015 dataset, but I do not understand which are the true that I should use to compute the correlation. The dataframe that you use contains several columns and I do not now which one should be the 'true' one. Also, I tried to compare the predictions I get with the results that you put in the file https://dl.fbaipublicfiles.com/fair-esm/examples/variant-prediction/data/raw_df.csv but there is no match. All I did is just running the script predict.py with scoring-strategy masked-marginals, which should be the best performing strategy according to the paper.

**Comment by nsh87** [comment]
@TheBil99 Errmm, I don’t think they’re different seeds. They pretrained ESM-1v on different subsets of UniRef, where sequences were clustered at different % identity levels to reduce the dataset size and stratify various taxa in each model. You can find these in UniClust. I believe the 5 models are 30%, 50%, 70%, 90% and 100% identity clustering for input sequences. They identify these in the paper. Is this not correct @tomsercu?

**Reply by tomsercu** [reply]
This is not correct! But this made me look back at [the paper](https://www.biorxiv.org/content/10.1101/2021.07.09.450648v2) so let's clarify:
* We did do an ablation studying the different clustering levels (Sec 4.2 "Pre-training data", Fig 4) - this motivates using Uniref90 throughout the paper. 
* We describe the Sec 4.1: "We use Uniref90" ... "We train five models with different seeds to produce an ensemble"

All that said ensembling across different pretrained models at different levels of clustering actually sounds like a pretty good idea :)

**Reply by nsh87** [reply]
Oh interesting! Heh, ya I would think various clust identities might generalize better across the TOL, but this generalizes fine already from what I’ve seen. Fascinating that there’s that much variability from the seeds - I suppose you could’ve trained a few more epochs to get rid of that, but then ESM-1v wouldn’t generalize as much. Was going to revisit the paper, thanks for this!
